lib/Common/Memory/RecyclerTelemetryInfo.cpp

```diff
@@ -0,0 +1,256 @@
+#endif
+
+    RecyclerTelemetryInfo::RecyclerTelemetryInfo(Recycler * recycler, RecyclerTelemetryHostInterface* hostInterface) :
+        passCount(0),
+        hostInterface(hostInterface),
+        lastPassStats(nullptr),
+        recyclerStartTime(Js::Tick::Now()),
+        captureTelemetryData(true)
+    {
+        this->recycler = recycler;
+
+        // temp
+        mainThreadID = GetCurrentThreadContextId();
+    }
+
+    const GUID& RecyclerTelemetryInfo::GetRecyclerID() const
+    {
+        return this->recycler->GetRecyclerID();
+    }
+
+    bool RecyclerTelemetryInfo::GetIsConcurrentEnabled() const
+    { 
+        return this->recycler->IsConcurrentEnabled(); 
+    }
+
+    RecyclerTelemetryInfo::~RecyclerTelemetryInfo()
+    {
+        AssertOnMainThread(this, RecyclerTelemetryInfo::~RecyclerTelemetryInfo);
+        this->FreeGCPassStats();
+    }
+
+    void RecyclerTelemetryInfo::FillInSizeData(IdleDecommitPageAllocator* allocator, AllocatorSizes* sizes) const
+    {
+        sizes->committedBytes = allocator->GetCommittedBytes();
+        sizes->reservedBytes = allocator->GetReservedBytes();
+        sizes->usedBytes = allocator->GetUsedBytes();
+        sizes->numberOfSegments = allocator->GetNumberOfSegments();
+    }
+
+    void RecyclerTelemetryInfo::StartPass()
+    {
+        Js::Tick start = Js::Tick::Now();
+        if(!this->captureTelemetryData)
+        { 
+            return;
+        }
+
+        //Output::Print(L"startPass called (current thread = %p, main thread = %p), passCount  = %i\r\n", GetCurrentThreadContextId(), this->mainThreadID, this->passCount);
+        AssertOnMainThread(this, RecyclerTelemetryInfo::StartPass);
+#if DBG
+        // validate state of existing GC pass stats structs
+        uint16 count = 0;
+        if (this->lastPassStats != nullptr)
+        {
+            RecyclerTelemetryGCPassStats* head = this->lastPassStats->next;
+            RecyclerTelemetryGCPassStats* curr = head;
+            do
+            {
+                AssertMsg(curr->isGCPassActive == false, "unexpected value for isGCPassActive");
+                count++;
+                curr = curr->next;
+            } while (curr != head);
+        }
+        AssertMsg(count == this->passCount, "RecyclerTelemetryInfo::StartPass() - mismatch between passCount and count.");
+#endif
+
+        RecyclerTelemetryGCPassStats* p = HeapNewNoThrow(RecyclerTelemetryGCPassStats);
+        if (p == nullptr)
+        {
+            // failed to allocate memory - disable any further telemetry capture for this recycler 
+            // and free any existing GC stats we've accumulated
+            this->captureTelemetryData = false;
+            FreeGCPassStats();
+            this->hostInterface->TransmitTelemetryError(*this, "Memory Allocation Failed");
+        }
+        else
+        {
+            passCount++;
+            memset(p, 0, sizeof(RecyclerTelemetryGCPassStats));
+            if (this->lastPassStats == nullptr)
```

Comment:
Curious if you could use SList here to avoid rolling your own linked list (and has the benefit of jd being able to traverse it easily)

Reply:
So, I should be able to do this, but I'm not immediately clear on how to get same allocator semantics that I have currently, which are "use the heap, don't throw".   I'd prefer to land this as-is at first & I'll modify to use SList in a follow-on change.

Reply:
//TODO added here.